src/dataflow/src/render/reduce.rs

```diff
@@ -31,23 +76,427 @@ use super::context::Context;
+                    full_aggrs,
+                    simple_aggrs,
+                    distinct_aggrs,
+                    num_accumulable,
+                })
+            }
+            ReductionType::Hierarchical => {
+                let aggr_funcs: Vec<_> = aggregates_list
+                    .iter()
+                    .cloned()
+                    .map(|(_, aggr)| aggr.func)
+                    .collect();
+                let indexes: Vec<_> = aggregates_list
+                    .into_iter()
+                    .map(|(index, _)| index)
+                    .collect();
+
+                // We don't have random access over Rows so we can simplify the
+                // task of grabbing the inputs we are aggregating over by
+                // generating a list of "skips" an iterator over the Row needs
+                // to do to get the desired indexes.
+                let skips = convert_indexes_to_skips(indexes);
+                if monotonic {
+                    let monotonic = MonotonicPlan { aggr_funcs, skips };
+                    ReducePlan::Hierarchical(HierarchicalPlan::Monotonic(monotonic))
+                } else {
+                    let mut buckets = vec![];
+                    let mut current = 16;
+
+                    // Plan for 4B records in the expected case if the user
+                    // didn't specify a group size.
+                    let limit = expected_group_size.unwrap_or(4_000_000_000);
+
+                    // Distribute buckets in powers of 16, so that we can strike
+                    // a balance between how many inputs each layer gets from
+                    // the preceeding layer, while also limiting the number of
+                    // layers.
+                    while current < limit {
+                        buckets.push(current as u64);
+                        current *= 16;
+                    }
+
+                    // We need to store the bucket numbers in decreasing order.
+                    buckets.reverse();
+
+                    let bucketed = BucketedPlan {
+                        aggr_funcs,
+                        skips,
+                        buckets,
+                    };
+
+                    ReducePlan::Hierarchical(HierarchicalPlan::Bucketed(bucketed))
+                }
+            }
+            ReductionType::Basic => {
+                if aggregates_list.len() == 1 {
+                    ReducePlan::Basic(BasicPlan::Single(
+                        aggregates_list[0].0,
+                        aggregates_list[0].1.clone(),
+                    ))
+                } else {
+                    ReducePlan::Basic(BasicPlan::Multiple(aggregates_list))
+                }
+            }
+        }
+    }
+
+    /// Generate a plan for computing the supplied aggregations.
+    ///
+    /// The resulting plan summarizes what the dataflow to be created
+    /// and how the aggregations will be executed.
+    fn create_from(
```

Comment:
Ah, I see. Lifting this up to be the first method makes the most sense to me (at least, as a person hoping to read top down).